packages/manager/src/managers/customTypes/CustomTypesManager.ts

```diff
@@ -237,49 +255,104 @@ export class CustomTypesManager extends BaseManager {
 
 					updateSharedSliceContentRelationships({
 						models: slices.models,
-						onUpdate: (model) => {
-							pushIfDefined(
-								crUpdates,
-								this.sliceMachinePluginRunner?.callHook("slice:update", {
-									libraryID: library.libraryID,
-									model,
-								}),
-							);
+						onUpdate: ({ previousModel, model }) => {
+							assertPluginsInitialized(this.sliceMachinePluginRunner);
+
+							crUpdates.push({
+								updatePromise: this.sliceMachinePluginRunner?.callHook(
+									"slice:update",
+									{ libraryID: library.libraryID, model },
+								),
+								rollback: () => {
+									this.sliceMachinePluginRunner?.callHook("slice:update", {
+										libraryID: library.libraryID,
+										model: previousModel,
+									});
+								},
+							});
 						},
 						previousPath,
 						newPath,
 					});
 				}
 
 				// Process all the Content Relationship updates at once.
-				const crUpdatesResult = await Promise.all(crUpdates);
+				const crUpdatesResult = await Promise.all(
+					crUpdates.map((update) => update.updatePromise),
+				);
 
 				if (crUpdatesResult.some((result) => result.errors.length > 0)) {
 					return {
 						errors: crUpdatesResult.flatMap((result) => result.errors),
+						rollback: async () => {
+							await Promise.all(crUpdates.map((update) => update.rollback()));
+						},
 					};
 				}
 			}
 		}
 
-		return { errors: [] };
+		return { errors: [], rollback: () => Promise.resolve() };
 	}
 
 	async updateCustomType(
 		args: CustomTypeUpdateHookData,
-	): Promise<OnlyHookErrors<CallHookReturnType<CustomTypeUpdateHook>>> {
+	): Promise<CustomTypesMachineManagerUpdateCustomTypeReturnType> {
 		assertPluginsInitialized(this.sliceMachinePluginRunner);
+		const { model } = args;
 
-		const hookResult = await this.sliceMachinePluginRunner.callHook(
+		let updateCrPromise: (() => Promise<{ errors: HookError[] }>) | undefined;
+
+		if (args.updateMeta?.fieldIdChanged) {
+			const customTypeRead = await this.readCustomType({ id: model.id });
+
+			if (customTypeRead.errors.length > 0) {
+				return { errors: customTypeRead.errors };
+			}
+			if (!customTypeRead.model) {
+				throw new Error(
+					"Read custom type without errors, but model is undefined.",
+				);
+			}
+
+			const previousCustomType = customTypeRead.model;
+
+			updateCrPromise = async () => {
+				const crUpdateResult = await this.updateContentRelationships(args);
+
+				if (crUpdateResult.errors.length > 0) {
```

Comment:
I don't get all this management.

I see a global top function that create a promise of update content relationship but still update the custom type passed in args that is then updated anyway from the main function with:

```
const customTypeUpdateResult = await this.sliceMachinePluginRunner.callHook(
			"custom-type:update",
			{ model },
		);
```

I think it could be a bit clearer, don't you think it's too complex here? Like you have some duplication? Something could be removed?

Reply:
This was done so that all the content relationship code was in the same block, IMO is better for readability. I changed to the previous approach so you can see if you find it better, I'm okay with both!